### PR TITLE
feat: create Segment analytics implementation + use geo-localization proxy (Batch 2 of 4)

### PR DESCRIPTION
**Pull request by NicolasMassart**
## **Description**

1. What is the reason for the change?
- Implement a new metrics system for the Segment SDK.
- Make it work alongside the old one so that we can remove the old one progressively.

3. What is the improvement/solution?
- Create a new class to handle metrics with segment
- abstract all segment SDK calls
- provide our own user and anonymous Ids for improved privacy
- Reuse the Mixpanel user id it available, otherwise generates a new one with the same format
- patch Segment SDK to allow the privacy aware user and anonymous Ids
- No event sent at all if user haven't accepted to send metrics info
- Total reset of the infos in case user asks to reset (in settings)
- Ability to disable all metrics
- Ability to ask for deletion of all data

## IMPORTANT

This PR contains a placeholder for the user metrics deletion request. This is to be implemented in a next PR in this batch but made fake here on purpose to prevent blocking this PR progress. See [the following placeholder code and comment in this PR.](https://github.com/MetaMask/metamask-mobile/pull/7520/files#r1411359734)

_NOTE: We discussed splitting this PR to extract the SDK patch in its own PR. But I would like to first try in only one PR, first because it's not so big, just 11 files, we've seen worse, but mostly because looking at the patch without the context of the MetaMetrics class seems strange and more complicated to me and could lead to mistakes or questions that are answered by the Metametrics code itself.
However, if you believe I should absolutely make it in a separate PR, I will do it. Thanks._

## **Related issues**

Fixes MetaMask/mobile-planning#1222

## **Manual testing steps**

Only unit tests, manual testing will be available for next step in this batch of work.

## **Screenshots/Recordings**

Bellow it no screenshot but the data received by Segment that we can see on the monitoring dev website.

image

### **Before**

NA

### **After**

Two event criteria:
- events being anonymous because user is not yet identified
- events being anonymous because even if user is identified we don't want to know
Example of anonymous event because user is not identified:
```json
{
  "context": {
    "library": {
      "name": "source-functions",
      "version": "1.0.0"
    },
    "location": {
      "api_processing_time_ms": 580,
      "country_code": "FR",
      "region": "NOR",
      "timezone": "Europe/Paris"
    }
  },
  "event": "183637normal anonymous event",
  "integrations": {},
  "messageId": "8f0e9b2e-a6ab-4b3f-922d-731932cce1a4",
  "originalTimestamp": "2023-11-16T17:36:45.43083525Z",
  "properties": {
    "applicationVersion": "7.10.0",
    "currentBuildNumber": "1187",
    "deviceBrand": "Apple",
    "operatingSystemVersion": "17.0.1",
    "platform": "ios"
  },
  "receivedAt": "2023-11-16T17:36:45.431Z",
  "sentAt": "2023-11-16T17:36:45.430Z",
  "timestamp": "2023-11-16T17:36:45.430Z",
  "type": "track",
  "userId": "0x0000000000000000",
  "writeKey": "REDACTED"
}
```
Example of anonymous event because user is not identified AND we don't want to know the details:
```json
{
  "context": {
    "library": {
      "name": "source-functions",
      "version": "1.0.0"
    },
    "location": {
      "api_processing_time_ms": 580,
      "country_code": "FR",
      "region": "NOR",
      "timezone": "Europe/Paris"
    }
  },
  "event": "183637anonymous anonymous event",
  "integrations": {},
  "messageId": "daa2a578-4778-49fb-91d2-389c864fb5e5",
  "originalTimestamp": "2023-11-16T17:36:45.430876388Z",
  "properties": {},
  "receivedAt": "2023-11-16T17:36:45.431Z",
  "sentAt": "2023-11-16T17:36:45.430Z",
  "timestamp": "2023-11-16T17:36:45.430Z",
  "type": "track",
  "userId": "0x0000000000000000",
  "writeKey": "REDACTED"
}
```

Then the user can identify:
```json
{
  "context": {
    "library": {
      "name": "source-functions",
      "version": "1.0.0"
    },
    "location": {
      "api_processing_time_ms": 0,
      "country_code": "FR",
      "region": "NOR",
      "timezone": "Europe/Paris"
    }
  },
  "integrations": {},
  "messageId": "72361f1f-3d8a-4b9a-97e1-7b01b457c4df",
  "originalTimestamp": "2023-11-16T17:37:13.309513172Z",
  "receivedAt": "2023-11-16T17:37:13.310Z",
  "sentAt": "2023-11-16T17:37:13.309Z",
  "timestamp": "2023-11-16T17:37:13.310Z",
  "traits": {
    "testtime": "183637"
  },
  "type": "identify",
  "userId": "0x7fb47bfe4e25c4e70097365f61838cfff275191e61f85262a51dc59fe52123de",
  "writeKey": "REDACTED"
}
```

and once identified, the events can still be sent anonymously or not depending if we have to know the details or not:

We know what but not who:
```json
{
  "context": {
    "library": {
      "name": "source-functions",
      "version": "1.0.0"
    },
    "location": {
      "api_processing_time_ms": 0,
      "country_code": "FR",
      "region": "NOR",
      "timezone": "Europe/Paris"
    }
  },
  "event": "183637delayed identified anonymous event",
  "integrations": {},
  "messageId": "4031820c-e2b0-4bfa-ae47-ccdf77a54d64",
  "originalTimestamp": "2023-11-16T17:37:13.30954482Z",
  "properties": {
    "applicationVersion": "7.10.0",
    "currentBuildNumber": "1187",
    "deviceBrand": "Apple",
    "operatingSystemVersion": "17.0.1",
    "platform": "ios"
  },
  "receivedAt": "2023-11-16T17:37:13.310Z",
  "sentAt": "2023-11-16T17:37:13.309Z",
  "timestamp": "2023-11-16T17:37:13.310Z",
  "type": "track",
  "userId": "0x0000000000000000",
  "writeKey": "REDACTED"
}
```

or we know who but not what:
```json
{
  "context": {
    "library": {
      "name": "source-functions",
      "version": "1.0.0"
    },
    "location": {
      "api_processing_time_ms": 0,
      "country_code": "FR",
      "region": "NOR",
      "timezone": "Europe/Paris"
    }
  },
  "event": "183637delayed identified anonymous event",
  "integrations": {},
  "messageId": "5076ef49-a852-4d23-8331-99ca81ee0e92",
  "originalTimestamp": "2023-11-16T17:37:13.309556282Z",
  "properties": {},
  "receivedAt": "2023-11-16T17:37:13.310Z",
  "sentAt": "2023-11-16T17:37:13.309Z",
  "timestamp": "2023-11-16T17:37:13.310Z",
  "type": "track",
  "userId": "0x7fb47bfe4e25c4e70097365f61838cfff275191e61f85262a51dc59fe52123de",
  "writeKey": "REDACTED"
}
```

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it is solved.
- [x] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [x] In case it's not yet "ready for review", I've set it to "draft".
  - [x] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Comment by NicolasMassart** [comment]
@SocketSecurity ignore @segment/tsub@2.0.0

I verified the said commit and it's just a package version fix

**Comment by NicolasMassart** [comment]
@SocketSecurity ignore-all
only stdlib version bump

File: app/core/Analytics/MetaMetrics.ts [hunk 1]
```diff
@@ -1,202 +1,211 @@
...
-              String(Math.round(Math.random() * Number.MAX_SAFE_INTEGER)),
-          ),
-        ),
-      );
-      await DefaultPreference.set(METAMETRICS_ID, metametricsId);
+    // look for a new Metametics ID and use it or generate a new one
+    this.metametricsId = await DefaultPreference.get(METAMETRICS_ID);
+    if (!this.metametricsId) {
+      // keep the id format compatible with MixPanel but base it on a UUIDv4
+      this.metametricsId = bufferToHex(keccak(uuidv4()));
```
Comment (NicolasMassart) [inline app/core/Analytics/MetaMetrics.ts:125]: C: this may change given UUID use here could be incompatible with the future Metamask Unified User Id. Discussion in progress, but consider this will just change the implementation of this specific part and doesn't makes the rest of the code irrelevant.

File: app/core/Analytics/MetaMetrics.ts [hunk 2]
```diff
@@ -230,85 +243,149 @@ class MetaMetrics implements IMetaMetrics {
...
+   * const metrics = await MetaMetrics.getInstance();
+   * ```
+   */
+  public static async getInstance(): Promise<IMetaMetrics> {
+    if (!this.instance) {
+      const config = {
         writeKey: (__DEV__
-          ? process.env.SEGMENT_DEV_KEY
-          : process.env.SEGMENT_PROD_KEY) as string,
+          ? process.env.SEGMENT_DEV_WRITE_KEY
```
Comment (sethkfman) [inline app/core/Analytics/MetaMetrics.ts:275]: To clarify this functionality.
- If you run in debug mode/locally will it use the dev segment environment?
- If you build QA/Prod on Bitrise will it use the PROD segment environment?

File: ios/Podfile.lock [hunk 3]
```diff
@@ -81,9 +81,8 @@ PODS:
   - lottie-react-native (5.1.5):
     - lottie-ios (~> 3.4.0)
     - React-Core
-  - MultiplatformBleAdapter (0.1.9)
   - OpenSSL-Universal (1.1.1100)
-  - Permission-BluetoothPeripheral (3.9.3):
+  - Permission-BluetoothPeripheral (3.10.1):
```
Comment (sethkfman) [inline ios/Podfile.lock:85]: I don't know if this value needs to be updated for this PR.

File: ios/Podfile.lock [hunk 4]
```diff
@@ -324,9 +323,6 @@ PODS:
     - React
   - react-native-background-timer (2.1.1):
     - React
-  - react-native-ble-plx (2.0.3):
```
Comment (sethkfman) [inline ios/Podfile.lock:327]: As well as removing this value

File: ios/Podfile.lock [hunk 5]
```diff
@@ -596,7 +592,6 @@ DEPENDENCIES:
   - React-logger (from `../node_modules/react-native/ReactCommon/logger`)
   - react-native-aes (from `../node_modules/react-native-aes-crypto`)
   - react-native-background-timer (from `../node_modules/react-native-background-timer`)
-  - react-native-ble-plx (from `../node_modules/react-native-ble-plx`)
```
Comment (sethkfman) [inline ios/Podfile.lock:599]: And removing these BLE values below.

File: patches/@segment+analytics-react-native+2.17.0.patch [hunk 6]
```diff
@@ -0,0 +1,150 @@
+diff --git a/node_modules/@segment/analytics-react-native/src/analytics.ts b/node_modules/@segment/analytics-react-native/src/analytics.ts
+index 79cc609..fff7c9b 100644
+--- a/node_modules/@segment/analytics-react-native/src/analytics.ts
++++ b/node_modules/@segment/analytics-react-native/src/analytics.ts
+@@ -62,6 +62,23 @@ import {
+ 
+ type OnPluginAddedCallback = (plugin: Plugin) => void;
+ 
++/**
++* ============================== PATCH INFORMATION ==============================
```
Comment (sethkfman) [inline patches/@segment+analytics-react-native+2.17.0.patch:10]: Great explanation

File: ios/Podfile.lock (hunk 5 above)
Comment (NicolasMassart) [reply ios/Podfile.lock:599]: These are some automatically updated dependencies side effect, will revert, sorry.

File: app/core/Analytics/MetaMetrics.ts [hunk 7]
```diff
@@ -208,107 +220,183 @@ class MetaMetrics implements IMetaMetrics {
...
+  //     await this.#storeDeleteRegulationCreationDate();
+  //     return { status: DataDeleteResponseStatus.ok };
+  //   } catch (error: any) {
+  //     Logger.error(error, 'Analytics Deletion Task Error');
+  //     return {
+  //       status: DataDeleteResponseStatus.error,
+  //       error: 'Analytics Deletion Task Error',
+  //     };
+  //   }
+  // };
```
Comment (NicolasMassart) [inline app/core/Analytics/MetaMetrics.ts:272]: In order to unblock this PR, this is to be implemented in another PR in the batch as we don't have the proxy yet.

File: app/core/Analytics/MetaMetrics.ts [hunk 8]
```diff
@@ -1,202 +1,213 @@
...
-              String(Math.round(Math.random() * Number.MAX_SAFE_INTEGER)),
-          ),
-        ),
-      );
-      await DefaultPreference.set(METAMETRICS_ID, metametricsId);
+    // look for a new Metametics ID and use it or generate a new one
+    this.metametricsId = await DefaultPreference.get(METAMETRICS_ID);
+    if (!this.metametricsId) {
+      // keep the id format compatible with MixPanel but base it on a UUIDv4
+      this.metametricsId = uuidv4();
```
Comment (NicolasMassart) [inline app/core/Analytics/MetaMetrics.ts:124]: It seems that Mixpanel uses UUIDv4 id already...
ie: F48D00D7-6A9F-43E7-9319-7C73CC567D7F
So attempting to be retrocompatible with a hex ethereum like address as id is probably a mistake.
And I assumed that keccak hashed UUIDv4 would make ids look like the current Mixpanel ones, but they don't.

So I will simply remove the keccak for now.
The upcoming ProfileID will probably be keccak hash format, but it will break compatibility in the usage anyway.
It will be less confusing if we use UUIDv4 for now so that it's clear when we change to the deterministic ProfileID that the id is different.
It will require updating existing user ids (alias?) but that's another story.

Note that I initially used a keccak formated ID because I was told to reuse the PR from January and this is what it used so I blindly assumed it was the right format.
It was probably a choice to match future format. But this is confusing and we should not plan code so far ahead in advance as it's clear in [YAGNI](https://martinfowler.com/bliki/Yagni.html):
> The first argument for yagni is that while we may now think we need this presumptive feature, it's likely that we will be wrong.

**Review by sethkfman — APPROVED** [review]
LGTM
